src/test/java/net/openhft/chronicle/core/time/UniqueMicroTimeProviderTest.java

```diff
@@ -19,59 +19,106 @@
 package net.openhft.chronicle.core.time;
 
 import net.openhft.chronicle.core.CoreTestCommon;
+import org.junit.Before;
 import org.junit.Test;
 
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertTrue;
 
 public class UniqueMicroTimeProviderTest extends CoreTestCommon {
+    private UniqueMicroTimeProvider timeProvider;
+    private SetTimeProvider setTimeProvider;
+
+    @Before
+    public void setUp() {
+        timeProvider = new UniqueMicroTimeProvider();
+        setTimeProvider = new SetTimeProvider(SystemTimeProvider.INSTANCE.currentTimeNanos());
+        timeProvider.provider(setTimeProvider);
+    }
 
     @Test
-    public void currentTimeMillis() throws IllegalStateException {
-        UniqueMicroTimeProvider tp = new UniqueMicroTimeProvider();
-        SetTimeProvider stp = new SetTimeProvider(SystemTimeProvider.INSTANCE.currentTimeNanos());
-        tp.provider(stp);
-        long last = 0;
-        for (int i = 0; i < 1_000; i++) {
-            stp.advanceNanos(i);
-            long time = tp.currentTimeMillis();
-            assertEquals(LongTime.toMicros(time), time);
-            assertTrue(time * 1000 > last);
-
-            long time2 = tp.currentTimeMicros();
-            assertTrue(time2 > last);
-
-            last = time2;
+    public void shouldProvideUniqueTimeAcrossThreads() throws InterruptedException {
+        final Set<Long> timeSet = ConcurrentHashMap.newKeySet();
+        final int numThreads = 100;
+        final int numIterations = 1_000;
+        ExecutorService executor = Executors.newFixedThreadPool(numThreads);
+        CountDownLatch latch = new CountDownLatch(numThreads);
+
+        for (int i = 0; i < numThreads; i++) {
+            executor.execute(() -> {
+                for (int j = 0; j < numIterations; j++) {
+                    setTimeProvider.advanceNanos(j);
+                    timeSet.add(timeProvider.currentTimeMicros());
```

Comment:
should also check that timestamps increase for this thread?

Reply:
I extended the test and checked for increasing values in a thread and unique values overall.